File: Python3/2462. Total Cost to Hire K Workers.py
class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        heap = []
        n = len(costs)

        for l in range(candidates):
            heapq.heappush(heap, (costs[l], 0))

        for r in range(max(candidates, n - candidates), n):   
            heapq.heappush(heap, (costs[r], 1))

        total = 0
        l, r = candidates, n - candidates - 1

        for _ in range(k):
            cand = heapq.heappop(heap)
            total += cand[0]
            if l > r:
                continue

            if cand[1] == 0:
                heapq.heappush(heap, (costs[l], 0))
                l += 1
            else:
                heapq.heappush(heap, (costs[r], 1))
                r -= 1
            
        return total

File: Java/2462. Total Cost to Hire K Workers.java
class Solution {
    public long totalCost(int[] costs, int k, int candidates) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0]) {
                return a[1] - b[1];
            }
            return a[0] - b[0];
        });

        for (int l = 0; l < candidates; l++) {
            pq.offer(new int[] {costs[l], 0});
        }

        int n = costs.length;
        for (int r = Math.max(candidates, n - candidates); r < n; r++) {
            pq.offer(new int[] {costs[r], 1});
        }

        long total = 0;
        int l = candidates, r = n - candidates - 1;

        for (int i = 0; i < k; i++) {
            int[] cur = pq.poll();
            total += cur[0];

            if (l <= r) {
                if (cur[1] == 0) {
                    pq.offer(new int[] {costs[l++], 0});
                } else {
                    pq.offer(new int[] {costs[r--], 1});
                }
            }
        }
        return total;
    }
}